src/ucs/sys/sock.c

```diff
@@ -951,6 +959,87 @@ int ucs_sockaddr_ip_cmp(const struct sockaddr *sa1, const struct sockaddr *sa2)
                   UCS_IPV4_ADDR_LEN : UCS_IPV6_ADDR_LEN);
 }
 
+static void ucs_sockaddr_print_subnet_info(const struct sockaddr *sa1,
+                                           const struct sockaddr *sa2,
+                                           unsigned prefix_len, int matched)
+{
+    UCS_STRING_BUFFER_ONSTACK(info, UCS_SOCKADDR_STRING_LEN * 2 + 60);
+    const struct sockaddr *saddrs[] = {sa1, sa2};
+    unsigned sa_index               = 0;
+    char ip_str[UCS_SOCKADDR_STRING_LEN];
+    const struct sockaddr **sa;
+    ucs_status_t status;
+
+    if (!ucs_log_is_enabled(UCS_LOG_LEVEL_DEBUG)) {
+        return;
+    }
+
+    ucs_string_buffer_appendf(&info, "IP addresses");
+    if (!matched) {
+        ucs_string_buffer_appendf(&info, " do not");
+    }
+
+    ucs_string_buffer_appendf(&info, " match with a %u-bit prefix,",
+                              prefix_len);
+
+    ucs_carray_for_each(sa, saddrs, ucs_static_array_size(saddrs)) {
+        status = ucs_sockaddr_get_ipstr(*sa, ip_str, UCS_SOCKADDR_STRING_LEN);
+        ucs_string_buffer_appendf(&info, " sa%u: %s", ++sa_index,
+                                  (status == UCS_OK) ? ip_str : "<null>");
+    }
+
+    ucs_debug("%s", ucs_string_buffer_cstr(&info));
+}
+
+ucs_status_t ucs_sockaddr_subnet_match(const struct sockaddr *sa1,
+                                       const struct sockaddr *sa2,
+                                       unsigned prefix_len, int *is_match_p)
+{
+    int matched         = 0;
+    ucs_status_t status = UCS_OK;
+    const void *ipaddr1, *ipaddr2;
+    size_t addr_size, addr_size_bits;
+
+    if (!ucs_sockaddr_are_both_known_af(sa1, sa2)) {
+        status = UCS_ERR_INVALID_PARAM;
```

Comment:
do we really need goto in this function? why not just return the status?

Reply:
won't it be problematic to add `*is_equal_p = is_equal;` before each return?

Reply:
Do you mean setting `is_match_p` (is_equal_p is from another function)?
Ok, I see, I thought it's not important to set that variable in case of an error, but if you think it's useful please keep the current impl